File: platform-server/app/schemas/job_execution.py
"""
Job Execution Tracking schemas
用于追踪 Job 执行情况的 schema 定义
"""
from datetime import datetime
from typing import Optional, List, Dict, Any

from pydantic import Field

from app.schemas.base import BaseSchema


class JobExecutionStats(BaseSchema):
    """Job 执行统计信息"""
    job_id: str = Field(..., description="Job ID")
    job_name: Optional[str] = Field(default=None, description="Job 名称")
    status: Optional[str] = Field(default=None, description="Job 状态")
    
    # 统计数据
    total_sub_jobs: int = Field(default=0, description="总 sub_job 数量")
    running_sub_jobs: int = Field(default=0, description="运行中的 sub_job 数量")
    completed_sub_jobs: int = Field(default=0, description="已完成的 sub_job 数量")
    failed_sub_jobs: int = Field(default=0, description="失败的 sub_job 数量")
    
    total_contents: int = Field(default=0, description="总文章数量")
    valid_contents: int = Field(default=0, description="有效文章数量")
    invalid_contents: int = Field(default=0, description="无效文章数量")
    pending_contents: int = Field(default=0, description="待确定文章数量")
    test_contents: int = Field(default=0, description="测试文章数量")
    
    # 目标与进度
    target_article_count: Optional[int] = Field(default=None, description="目标文章数")
    progress_percentage: Optional[float] = Field(default=None, description="完成百分比")
    
    # 时间信息
    first_sub_job_time: Optional[datetime] = Field(default=None, description="首个 sub_job 创建时间")
    last_sub_job_time: Optional[datetime] = Field(default=None, description="最新 sub_job 创建时间")
    
    class Config:
        from_attributes = True


class SubJobDetail(BaseSchema):
    """SubJob 详情（包含关联的 content 信息）"""
    id: int
    job_id: str
    sub_job_id: str
    content_id: str
    expert_list: List[str]
    expert_complete_list: Optional[List[str]] = None
    status: str
    error_message: Optional[str] = None
    create_time: Optional[datetime] = None
    update_time: Optional[datetime] = None
    plan_index: Optional[int] = Field(default=None, description="执行计划索引")

    # 关联的 content 信息
    content_title: Optional[str] = Field(default=None, description="文章标题")
    content_is_valid: Optional[int] = Field(default=None, description="文章是否有效")
    content_is_test: Optional[int] = Field(default=None, description="是否测试文章")
    
    class Config:
        from_attributes = True


class CriticScoreItem(BaseSchema):
    """单个专家的评分详情"""
    expert_func: str = Field(..., description="专家函数名")
    expert_type: str = Field(default="CRITIC", description="专家类型：BAN/CRITIC")
    score: int = Field(..., description="评分 0-100")
    passed: bool = Field(..., description="是否通过")
    reason: Optional[str] = Field(default=None, description="评分理由")


class ContentCriticSummary(BaseSchema):
    """Critic 评分汇总"""
    total_critics: int = Field(default=0, description="Critic 专家数量")
    passed_count: int = Field(default=0, description="通过数量")
    failed_count: int = Field(default=0, description="不通过数量")
    avg_score: Optional[float] = Field(default=None, description="平均分")
    min_score: Optional[int] = Field(default=None, description="最低分")
    has_ban_issue: bool = Field(default=False, description="是否有合规问题（BAN 类专家不通过）")
    problem_count: int = Field(default=0, description="问题总数")
    # 每个专家的评分详情
    scores: List[CriticScoreItem] = Field(default_factory=list, description="各专家评分详情")


class ContentDetail(BaseSchema):
    """Content 详情"""
    id: int
    job_id: str
    sub_job_id: str
    content_id: str
    agent_code: Optional[str] = Field(None, description="Agent 编码")
    prompt: Optional[str] = None
    context_list: Optional[Dict[str, str]] = Field(
        default=None,
        description="结构化变量映射，格式: {变量名: 上下文名}"
    )
    title: Optional[str] = None
    content: Optional[str] = None
    is_valid: Optional[int] = None
    is_test_case: int
    online_status: Optional[str] = Field(default="OFFLINE", description="上线状态")
    is_locked: int = Field(default=0, description="是否锁定：0否 1是")
    is_used: int = Field(default=0, description="是否被使用：0否 1是")
    create_time: Optional[datetime] = None
    update_time: Optional[datetime] = None

    # SubJob 状态
    sub_job_status: Optional[str] = Field(default=None, description="关联的 sub_job 状态")

    # Critic 评分汇总
    critic_summary: Optional[ContentCriticSummary] = Field(default=None, description="Critic 评分汇总")

    class Config:
        from_attributes = True


class JobExecutionDetail(BaseSchema):
    """Job 执行详情（包含统计、sub_job 列表、content 列表）"""
    stats: JobExecutionStats
    sub_jobs: List[SubJobDetail] = Field(default_factory=list, description="SubJob 列表")
    contents: List[ContentDetail] = Field(default_factory=list, description="Content 列表")


class ExpertBusinessResultDetail(BaseSchema):
    """Expert 业务结果详情"""
    id: int
    job_id: str
    sub_job_id: str
    content_id: str
    expert_task_id: Optional[int] = None
    expert_config_code: str
    expert_config_name: Optional[str] = None
    expert_func: Optional[str] = None  # 专家函数名（CriticIllegal 等）
    expert_type: Optional[str] = None  # 专家类型：BAN/CRITIC
    model_code: Optional[str] = None
    business_type: Optional[str] = None
    plugin_config_snapshot: Optional[List[Any]] = None
    prompt: Optional[str] = None
    business_result: Optional[Any] = None
    status: Optional[str] = None
    error_message: Optional[str] = None
    create_time: Optional[datetime] = None
    
    class Config:
        from_attributes = True

File: platform-console/apps/raap-admin/src/api/core/job-execution.ts
/**
 * Job 执行追踪相关 API
 */
import { requestClient } from '#/api/request';

/**
 * Job 执行统计信息
 */
export interface JobExecutionStats {
  job_id: string;
  job_name?: string;
  status?: string;
  // 统计数据
  total_sub_jobs: number;
  running_sub_jobs: number;
  completed_sub_jobs: number;
  failed_sub_jobs: number;
  total_contents: number;
  valid_contents: number;
  invalid_contents: number;
  test_contents: number;
  // 目标与进度
  target_article_count?: number;
  progress_percentage?: number;
  // 时间信息
  first_sub_job_time?: string;
  last_sub_job_time?: string;
}

/**
 * SubJob 详情
 */
export interface SubJobDetail {
  id: number;
  job_id: string;
  sub_job_id: string;
  content_id: string;
  expert_list: string[];
  expert_complete_list?: string[];
  status: string;
  error_message?: string;
  create_time?: string;
  update_time?: string;
  // 关联的 content 信息
  content_title?: string;
  content_is_valid?: number;
  content_is_test?: number;
}

/**
 * 单个专家的评分详情
 */
export interface CriticScoreItem {
  /** 专家函数名 */
  expert_func: string;
  /** 专家类型：BAN/CRITIC */
  expert_type: string;
  /** 评分 0-100 */
  score: number;
  /** 是否通过 */
  passed: boolean;
  /** 评分理由 */
  reason?: string;
}

/**
 * Critic 评分汇总
 */
export interface ContentCriticSummary {
  /** Critic 专家数量 */
  total_critics: number;
  /** 通过数量 */
  passed_count: number;
  /** 不通过数量 */
  failed_count: number;
  /** 平均分 */
  avg_score: null | number;
  /** 最低分 */
  min_score: null | number;
  /** 是否有合规问题（BAN 类专家不通过） */
  has_ban_issue: boolean;
  /** 问题总数 */
  problem_count: number;
  /** 各专家评分详情 */
  scores: CriticScoreItem[];
}

/**
 * Content 详情
 */
export interface ContentDetail {
  id: number;
  job_id: string;
  sub_job_id: string;
  content_id: string;
  agent_code?: string;
  prompt?: string;
  /**
   * 兼容两种结构：
   * - 旧：string[]
   * - 新：Record<key, value>
   */
  context_list?: Record<string, unknown> | string[];
  title?: string;
  content?: string;
  is_valid: number;
  is_test_case: number;
  online_status?: 'OFFLINE' | 'ONLINE';
  is_locked: number;
  is_used: number;
  create_time?: string;
  update_time?: string;
  // SubJob 状态
  sub_job_status?: string;
  // Critic 评分汇总
  critic_summary?: ContentCriticSummary;
}

/**
 * Expert 业务结果详情
 */
export interface ExpertBusinessResultDetail {
  id: number;
  job_id: string;
  sub_job_id: string;
  content_id: string;
  expert_task_id?: number;
  expert_config_code: string;
  expert_config_name?: string;
  /** 专家函数名（CriticIllegal 等） */
  expert_func?: string;
  /** 专家类型：BAN/CRITIC */
  expert_type?: string;
  model_code?: string;
  business_type?: string;
  plugin_config_snapshot?: unknown[];
  prompt?: string;
  business_result?: Record<string, unknown>;
  status?: string;
  error_message?: string;
  create_time?: string;
}

/**
 * Job 执行详情
 */
export interface JobExecutionDetail {
  stats: JobExecutionStats;
  sub_jobs: SubJobDetail[];
  contents: ContentDetail[];
}

/**
 * Job 执行统计列表响应
 */
export interface JobExecutionStatsListResponse {
  total: number;
  items: JobExecutionStats[];
}

/**
 * 获取所有 Job 执行统计
 */
export async function getAllJobExecutionStatsApi(params?: {
  page?: number;
  page_size?: number;
}) {
  return requestClient.get<JobExecutionStatsListResponse>(
    '/v1/job-execution/stats',
    {
      params,
    },
  );
}

/**
 * 获取指定 Job 执行统计
 */
export async function getJobExecutionStatsApi(jobId: string) {
  return requestClient.get<JobExecutionStats>(
    `/v1/job-execution/${jobId}/stats`,
  );
}

/**
 * 获取指定 Job 执行详情
 */
export async function getJobExecutionDetailApi(
  jobId: string,
  params?: { include_test?: boolean },
) {
  return requestClient.get<JobExecutionDetail>(
    `/v1/job-execution/${jobId}/detail`,
    { params },
  );
}

/**
 * 获取指定 Job 的 SubJob 列表
 */
export async function getJobSubJobsApi(
  jobId: string,
  params?: {
    is_test_case?: number;
    is_valid?: number;
    limit?: number;
    skip?: number;
    status?: string;
  },
) {
  return requestClient.get<SubJobDetail[]>(
    `/v1/job-execution/${jobId}/sub-jobs`,
    {
      params,
    },
  );
}

/**
 * 获取指定 Job 的 Content 列表
 */
export async function getJobContentsApi(
  jobId: string,
  params?: {
    is_test_case?: number;
    is_valid?: number;
    limit?: number;
    skip?: number;
  },
) {
  return requestClient.get<ContentDetail[]>(
    `/v1/job-execution/${jobId}/contents`,
    {
      params,
    },
  );
}

/**
 * 获取指定 Job 的 Expert 业务结果列表
 */
export async function getJobBusinessResultsApi(
  jobId: string,
  params?: {
    content_id?: string;
    expert_config_code?: string;
    limit?: number;
    skip?: number;
  },
) {
  return requestClient.get<ExpertBusinessResultDetail[]>(
    `/v1/job-execution/${jobId}/business-results`,
    { params },
  );
}
